crates/api_common/src/utils.rs

```diff
@@ -126,52 +132,67 @@ pub async fn mark_post_as_unread(
     .with_lemmy_type(LemmyErrorType::CouldntMarkPostAsRead)
 }
 
-pub fn check_user_valid(
-  banned: bool,
-  ban_expires: Option<DateTime<Utc>>,
-  deleted: bool,
-) -> Result<(), LemmyError> {
+pub fn check_user_valid(person: &Person) -> Result<(), LemmyError> {
   // Check for a site ban
-  if is_banned(banned, ban_expires) {
+  if is_banned(person.banned, person.ban_expires) {
     Err(LemmyErrorType::SiteBan)?
   }
   // check for account deletion
-  else if deleted {
+  else if person.deleted {
     Err(LemmyErrorType::Deleted)?
   } else {
     Ok(())
   }
 }
 
-#[tracing::instrument(skip_all)]
-pub async fn check_community_ban(
-  person_id: PersonId,
+pub async fn check_community_action(
+  person: &Person,
   community_id: CommunityId,
   pool: &mut DbPool<'_>,
-) -> Result<(), LemmyError> {
-  let is_banned = CommunityPersonBanView::get(pool, person_id, community_id)
+) -> LemmyResult<()> {
+  check_user_valid(person)?;
+
+  // check if community was deleted or removed
+  let community = Community::read(pool, community_id)
+    .await
+    .with_lemmy_type(LemmyErrorType::CouldntFindCommunity)?;
+  if community.deleted || community.removed {
+    Err(LemmyErrorType::Deleted)?
+  }
+
+  // check if user was banned from site or community
+  // TODO: this reads unnecessary data in case of ban (both person and community)
+  let is_banned = CommunityPersonBanView::get(pool, person.id, community_id)
     .await
     .is_ok();
   if is_banned {
     Err(LemmyErrorType::BannedFromCommunity)?
-  } else {
-    Ok(())
   }
+
+  Ok(())
 }
 
-#[tracing::instrument(skip_all)]
-pub async fn check_community_deleted_or_removed(
+pub async fn check_community_mod_action(
+  person: &Person,
   community_id: CommunityId,
   pool: &mut DbPool<'_>,
-) -> Result<(), LemmyError> {
-  let community = Community::read(pool, community_id)
-    .await
-    .with_lemmy_type(LemmyErrorType::CouldntFindCommunity)?;
-  if community.deleted || community.removed {
-    Err(LemmyErrorType::Deleted)?
+) -> LemmyResult<()> {
+  check_community_action(person, community_id, pool).await?;
+  is_mod_or_admin(pool, person, community_id).await?;
+  Ok(())
+}
+
+pub async fn check_community_mod_action_opt(
+  local_user_view: &LocalUserView,
+  community_id: Option<CommunityId>,
+  pool: &mut DbPool<'_>,
+) -> LemmyResult<()> {
+  if let Some(community_id) = community_id {
+    check_community_mod_action(&local_user_view.person, community_id, pool).await?;
   } else {
-    Ok(())
+    is_admin(local_user_view)?;
   }
+  Ok(())
 }
```

Comment:
Main changes and cleanup in this file. I noticed that some of the DB views read unnecessary data (see todos).